src/colvarcomp_rotations.cpp

```diff
@@ -131,9 +127,13 @@ void colvar::orientation::apply_force(colvarvalue const &force)
   cvm::quaternion const &FQ = force.quaternion_value;
 
   if (!atoms->noforce) {
+    rotation_derivative<cvm::atom_pos, cvm::atom_pos> rot_deriv(rot, ref_pos, shifted_pos);
+    rot_deriv.prepare_derivative(false, true);
```

Comment:
The term "prepare" sounds ambiguous between "initialize" or "compute", but that's more or less what that function does.

However, of the two arguments here the second `require_dq` is always true, but the first one `require_dl` is used only during debug. Perhaps their order could be switched and they should be optional arguments?

Also, a further simplification could be to incorporate that call directly in `rotation::calc_optimal_rotation_impl()`? Since it's idempotent the debug code could simply call it again with the right flags.

Reply:
I have used a `enum class rotation_derivative_dldq` to make the option clearer. `dq` is for the derivative of the leading eigenvector with respect to the atomic coordinates, and `dl` is for the derivative of the leading eigenvalue with respect to the atomic coordinates. Although current code only requires `dq` except `debug_gradients`, there could be special CVs only depending on `dl`, so making the parameter default to `dq` may not be the best idea.

The idea of this PR is avoiding unnecessary calculation in `rotation::calc_optimal_rotation` as much as possible, and if we put `prepare_derivative` back to `calc_optimal_rotation_impl`, then each time when we calculate the optimal rotation, we will calculate the temporary variables in `prepare_derivative`. There are two cases where the temporary variables are not necessary:

1. RMSD. RMSD does not require `dq` or `dl` even in `calc_gradients()`, so if `prepare_derivative` is called in `calc_optimal_rotation_impl`, some typical use cases of RMSD, such as aligning structures by employing TMD along RMSD, may slow down;
2. The derivative is not required at all due to no biasing force. For example, if we just want to monitor the values of CVs during simulations, without applying any biases or restraints, then all `calc_gradients()` are not called, and `prepare_derivative` should be avoided in this case.

Reply:
> The idea of this PR is avoiding unnecessary calculation in `rotation::calc_optimal_rotation` as much as possible, and if we put `prepare_derivative` back to `calc_optimal_rotation_impl`, then each time when we calculate the optimal rotation, we will calculate the temporary variables in `prepare_derivative`.

Correct, but in the same way as when flags are passed at compute time the `calc_optimal_rotation_impl` object can also be told from the beginning to do only the work that's needed. 

At any rate, that code path will likely undergo more revisions, so it's not worth spending more effort there. As long as there is a performance improvement and readability doesn't get significantly worse, it's all great.

On that note, are you okay with [06e61c0](https://github.com/Colvars/colvars/pull/541/commits/06e61c084d8a8a0d567d26f03c08b7552af8360b) (i.e. casting to bool when you are clearly using the bitwise "and" operator in conditionals)?

Reply:
Thanks! I am OK with it.